### PR TITLE
[experimental] Add multinode profiling workflow

### PR DESCRIPTION
**Pull request by hbarclay**

### CODE REVIEW

File: patches/trtllm-profiling.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        print(f"[trtllm-patch] ERROR: HandlerBase has no generate_locally. "
+              f"Available: {methods}", file=sys.stderr, flush=True)
+        return
+
+    os.makedirs(output_dir, exist_ok=True)
+
+    _orig_generate_locally = HandlerBase.generate_locally
+    _state = {"step": 0, "started": False, "stopped": False}
+
+    _state["profiler"] = torch.profiler.profile(
```
Comment (briancoutinho) [inline patches/trtllm-profiling.py]: One idea is you can also consider adding a schedule with start /stop step and 1 warmup iteratipm
https://docs.pytorch.org/docs/stable/profiler.html#torch.profiler.schedule

Reference in Megatron Core
https://github.com/NVIDIA/Megatron-LM/blob/ab43d43f0bc04f4656d4af15afb6e7e4c9ad71c8/megatron/training/training.py#L2846-L2861

File: patches/trtllm-profiling.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        activities=[
+            torch.profiler.ProfilerActivity.CPU,
+            torch.profiler.ProfilerActivity.CUDA,
+        ],
+        record_shapes=True,
+        with_stack=False,
+        on_trace_ready=torch.profiler.tensorboard_trace_handler(output_dir),
+    )
+
+    async def _patched_generate_locally(self, request, context, embeddings=None):
```
Comment (briancoutinho) [inline patches/trtllm-profiling.py:46]: If you use torch profiler with schedule you can add a call to
_state["profiler"].step() on every iteration.

This allows you to configure the start and stop automatically.
It will also add a "Step<N>" marker in the trace that might be useful

File: patches/trtllm-profiling.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    _state = {"step": 0, "started": False, "stopped": False}
+
+    _state["profiler"] = torch.profiler.profile(
+        activities=[
+            torch.profiler.ProfilerActivity.CPU,
+            torch.profiler.ProfilerActivity.CUDA,
+        ],
+        record_shapes=True,
+        with_stack=False,
+        on_trace_ready=torch.profiler.tensorboard_trace_handler(output_dir),
```
Comment (briancoutinho) [inline patches/trtllm-profiling.py]: Not important but I typically pass in my own handler so I can control the file naming, default one does not include rank in the name.
See this reference in Megatron Core
https://github.com/NVIDIA/Megatron-LM/blob/ab43d43f0bc04f4656d4af15afb6e7e4c9ad71c8/megatron/training/training.py#L2846-L2861

File: patches/trtllm-profiling.py (hunk 2 above)
Comment (briancoutinho) [reply patches/trtllm-profiling.py:46]: https://github.com/NVIDIA/Megatron-LM/blob/ab43d43f0bc04f4656d4af15afb6e7e4c9ad71c8/megatron/training/training.py#L2899-L2900

**Review by briancoutinho — COMMENTED** [review]
Added some suggestions for the profiler patch, lgtm otherwise

**Comment by cquil11** [comment]
Closing due to inactivity.
